### PR TITLE
feat: qos queue trimming

### PR DESCRIPTION
**Pull request by denopink**
- 100% qos test cover
- implement `priorityQueue.trim()` to remove messages with the lowest qos value while honoring the configured qos `priority` [newest, oldest message]
- the compute cost of trim will be O(n)

### CODE REVIEW

File: internal/wrphandlers/qos/options.go [hunk 1]
```diff
@@ -47,25 +58,83 @@ func MaxMessageBytes(s int) Option {
 		})
 }
 
-// Priority determines what is used [newest, oldest message] for QualityOfService tie breakers,
+// Priority determines what is used [newest, oldest message] for QualityOfService tie breakers and trimming,
 // with the default being to prioritize the newest messages.
 func Priority(p PriorityType) Option {
```
Comment (denopink) [inline internal/wrphandlers/qos/options.go:58]: needed to improve test coverage

File: internal/wrphandlers/qos/qos.go [hunk 2]
```diff
@@ -55,13 +66,17 @@ func New(next wrpkit.Handler, opts ...Option) (*Handler, error) {
 		return nil, ErrInvalidInput
 	}
 
-	// Add configuration validators.
-	opts = append(opts, validateQueueConstraints(), validatePriority(), validateTieBreaker())
-
 	h := Handler{
```
Comment (denopink) [inline internal/wrphandlers/qos/qos.go:72]: keeping how we assign defaults consistent

File: internal/wrphandlers/qos/priority_queue.go [hunk 3]
```diff
@@ -28,24 +32,44 @@ type priorityQueue struct {
...
+		ok  bool
+	)
+	for pq.Len() != 0 {
+		itm := heap.Pop(pq).(item)
+		// itm.expired will be true if `itm` has already been `pq.trim()'.
+		// If itm.expired is true, then pop the next item.
+		if !itm.expired {
+			msg = *itm.msg
+			ok = true
+			break
```
Comment (denopink) [inline internal/wrphandlers/qos/priority_queue.go]: removing expired items as we `heap.Pop()` saves us a `heap.Init()`

File: internal/wrphandlers/qos/priority_queue.go [hunk 4]
```diff
@@ -61,17 +85,73 @@ func (pq *priorityQueue) Enqueue(msg wrp.Message) error {
...
-	for pq.sizeBytes > pq.maxQueueBytes {
-		// Note, priorityQueue.drop does not drop the least prioritized queued message.
-		// i.e.: a high priority queued message may be dropped instead of a lesser queued message.
-		pq.drop()
+	// If priorityQueue.queue doesn't violates `maxQueueSize`, then return.
+	if pq.sizeBytes <= pq.maxQueueBytes {
+		return
+	}
+
+	itemsCache := make([]*item, len(pq.queue))
```
Comment (denopink) [inline internal/wrphandlers/qos/priority_queue.go:102]: `itemsCache []*item` saves us from making unnecessary copies

File: internal/wrphandlers/qos/priority_queue.go [hunk 5]
```diff
@@ -61,17 +85,73 @@ func (pq *priorityQueue) Enqueue(msg wrp.Message) error {
...
+			continue
+		}
+
+		itemsCache[iCache] = itm
+		iCache += 1
+	}
+
+	// Resize itemsCache.
+	itemsCache = itemsCache[:iCache]
+	slices.SortFunc(itemsCache, func(i, j *item) int {
```
Comment (denopink) [inline internal/wrphandlers/qos/priority_queue.go:129]: `SortFunc` sorts the slice in ascending order based on `func(i, j)`

File: internal/wrphandlers/qos/options.go [hunk 6]
```diff
@@ -6,11 +6,19 @@ package qos
...
 const (
+	// Priority queue defaults.
 	DefaultMaxQueueBytes   = 1 * 1024 * 1024 // 1MB max/queue
 	DefaultMaxMessageBytes = 256 * 1024      // 256 KB
+
+	// QOS expires defaults.
+	DefaultLowQOSExpires      = time.Minute * 15
+	DefaultMediumQOSExpires   = time.Minute * 20
+	DefaultHighQOSExpires     = time.Minute * 25
+	DefaultCriticalQOSExpires = time.Minute * 30
```
Comment (denopink) [inline internal/wrphandlers/qos/options.go]: parodus defaults

File: internal/wrphandlers/qos/priority_queue.go [hunk 7]
```diff
@@ -28,24 +32,44 @@ type priorityQueue struct {
...
+
+	// QOS expiries.
+	// lowQOSExpires determines when low qos messages are trimmed.
+	lowQOSExpires time.Duration
+	// mediumQOSExpires determines when medium qos messages are trimmed.
+	mediumQOSExpires time.Duration
+	// highQOSExpires determines when high qos messages are trimmed.
+	highQOSExpires time.Duration
+	// criticalQOSExpires determines when critical qos messages are trimmed.
+	criticalQOSExpires time.Duration
```
Comment (denopink) [inline internal/wrphandlers/qos/priority_queue.go]: configurable

File: internal/wrphandlers/qos/priority_queue.go [hunk 8]
```diff
@@ -19,6 +20,9 @@ var ErrMaxMessageBytes = errors.New("wrp message payload exceeds maxMessageBytes
 type priorityQueue struct {
 	// queue for wrp messages, ingested by serviceQOS
 	queue []item
+	// Priority determines what is used [newest, oldest message] for QualityOfService tie breakers and trimming,
+	// with the default being to prioritize the newest messages.
+	priority PriorityType
```
Comment (denopink) [inline internal/wrphandlers/qos/priority_queue.go:25]: already configurable

File: cmd/xmidt-agent/config.go [hunk 9]
```diff
@@ -64,9 +64,17 @@ type QOS struct {
 	MaxQueueBytes int64
 	// MaxMessageBytes is the largest allowable wrp message payload.
 	MaxMessageBytes int
-	// Priority determines what is used [newest, oldest message] for QualityOfService tie breakers,
+	// Priority determines what is used [newest, oldest message] for QualityOfService tie breakers and trimming,
 	// with the default being to prioritize the newest messages.
 	Priority qos.PriorityType
+	// LowQOSExpires determines when low qos messages are trimmed.
+	LowQOSExpires time.Duration
```
Comment (johnabass) [inline cmd/xmidt-agent/config.go]: You don't need 'QOS' in the names of these settings, as that's already in the name of the enclosing struct

Comment (denopink) [reply cmd/xmidt-agent/config.go]: sounds good!

**Review by johnabass — APPROVED** [review]
LGTM!

**Pull request merged**